File: util-validator/src/test/scala/com/twitter/util/validation/internal/validators/ValidPassengerCountConstraintValidator.scala
package com.twitter.util.validation.internal.validators

import com.twitter.util.validation.caseclasses.Car
import com.twitter.util.validation.constraints.ValidPassengerCount
import jakarta.validation.{ConstraintValidator, ConstraintValidatorContext}

class ValidPassengerCountConstraintValidator extends ConstraintValidator[ValidPassengerCount, Car] {

  @volatile private[this] var maxPassengers: Long = _

  override def initialize(constraintAnnotation: ValidPassengerCount): Unit = {
    this.maxPassengers = constraintAnnotation.max()
  }

  override def isValid(
    car: Car,
    constraintValidatorContext: ConstraintValidatorContext
  ): Boolean =
    if (car == null) true
    else car.passengers.nonEmpty && car.passengers.size <= this.maxPassengers
}

File: util-validator/src/test/java/com/twitter/util/validation/tests/ValidPassengerCountConstraintValidator.java
package com.twitter.util.validation.tests;

import com.twitter.util.validation.constraints.ValidPassengerCount;

import jakarta.validation.ConstraintValidator;
import jakarta.validation.ConstraintValidatorContext;

public class ValidPassengerCountConstraintValidator implements ConstraintValidator<ValidPassengerCount, Car> {

  private long max;

  @Override
  public void initialize(ValidPassengerCount constraintAnnotation) {
    this.max = constraintAnnotation.max();
  }

  @Override
  public boolean isValid(Car car, ConstraintValidatorContext context) {
    if ( car == null ) {
      return true;
    }

    return car.getPassengers().size() <= this.max;
  }
}
